File: src/game/player.ts
import type { WebSocket } from 'ws';
import type { CombatSnapshot, PlayerSnapshot } from '../protocol/messages.js';
import type { StoredPlayer } from '../db/database.js';
import { ITEMS } from './items.js';
import { CLASSES, xpForLevel } from './types.js';
import type { World } from './world.js';
import { findGuildByMember } from './guilds.js';
import type { QuestProgress } from './quests.js';

export class PlayerSession {
  readonly ws: WebSocket;
  data: StoredPlayer;
  combatTarget: string | null = null;
  pvpTarget: string | null = null;
  inDuel = false;
  partyLeader: string | null = null;
  searchedRooms = new Set<string>();
  authenticated = false;

  get inCombat(): boolean {
    return this.combatTarget !== null || this.pvpTarget !== null;
  }

  clearCombat(): void {
    this.combatTarget = null;
    this.pvpTarget = null;
  }

  constructor(ws: WebSocket, data: StoredPlayer) {
    this.ws = ws;
    this.data = data;
  }

  get username(): string {
    return this.data.username;
  }

  get roomId(): string {
    return this.data.roomId;
  }

  get totalAttack(): number {
    const cls = CLASSES[this.data.className];
    let atk = cls.attack + (this.data.level - 1) * 2;
    if (this.data.equipment.weapon) {
      const weapon = ITEMS[this.data.equipment.weapon];
      if (weapon?.attack) atk += weapon.attack;
    }
    return atk;
  }

  get totalDefense(): number {
    const cls = CLASSES[this.data.className];
    let def = cls.defense + (this.data.level - 1);
    if (this.data.equipment.armor) {
      const armor = ITEMS[this.data.equipment.armor];
      if (armor?.defense) def += armor.defense;
    }
    return def;
  }

  get xpToLevel(): number {
    return xpForLevel(this.data.level);
  }

  addXp(amount: number): string[] {
    const messages: string[] = [];
    this.data.xp += amount;
    messages.push(`You gain ${amount} experience.`);

    while (this.data.xp >= this.xpToLevel) {
      this.data.xp -= this.xpToLevel;
      this.levelUp(messages);
    }
    return messages;
  }

  private levelUp(messages: string[]): void {
    this.data.level++;
    const cls = CLASSES[this.data.className];
    const hpGain = Math.floor(cls.maxHp * 0.15) + 5;
    const mpGain = Math.floor(cls.maxMp * 0.12) + 3;
    this.data.maxHp += hpGain;
    this.data.maxMp += mpGain;
    this.data.hp = this.data.maxHp;
    this.data.mp = this.data.maxMp;
    messages.push(`*** LEVEL UP! You are now level ${this.data.level}! ***`);
    messages.push(`HP +${hpGain}, MP +${mpGain}`);
  }

  addItem(itemId: string): void {
    this.data.inventory.push(itemId);
  }

  removeItem(itemId: string): boolean {
    const idx = this.data.inventory.indexOf(itemId);
    if (idx === -1) return false;
    this.data.inventory.splice(idx, 1);
    return true;
  }

  countItem(itemId: string): number {
    return this.data.inventory.filter((id) => id === itemId).length;
  }

  getActiveQuests(): QuestProgress[] {
    return this.data.quests.filter((q) => q.status === 'active');
  }

  toSnapshot(world?: World): PlayerSnapshot {
    const room = world?.rooms.get(this.data.roomId);
    return {
      username: this.data.username,
      className: this.data.className,
      level: this.data.level,
      hp: this.data.hp,
      maxHp: this.data.maxHp,
      mp: this.data.mp,
      maxMp: this.data.maxMp,
      xp: this.data.xp,
      xpToLevel: this.xpToLevel,
      gold: this.data.gold,
      room: this.data.roomId,
      roomName: room?.name,
      zone: room?.zone,
      inCombat: this.inCombat,
      partyLeader: this.partyLeader ?? undefined,
      inDuel: this.inDuel,
      title: this.data.title ?? undefined,
      guildName: findGuildByMember(this.username)?.name,
    };
  }

  toCombatSnapshot(): CombatSnapshot {
    return {
      inCombat: this.inCombat,
      target: this.pvpTarget ?? this.combatTarget ?? undefined,
    };
  }
}

File: crates/realm-core/src/player.rs
use crate::db::StoredPlayer;
use crate::guilds::find_guild_by_member;
use crate::items::ITEMS;
use crate::quests::{QuestProgress, QuestStatus};
use crate::types::{class_stats, xp_for_level};
use crate::world::World;
use realm_protocol::PlayerSnapshot;
use std::collections::HashSet;

#[derive(Clone)]
pub struct PlayerSession {
    pub data: StoredPlayer,
    pub combat_target: Option<String>,
    pub pvp_target: Option<String>,
    pub in_duel: bool,
    pub party_leader: Option<String>,
    pub searched_rooms: HashSet<String>,
    pub authenticated: bool,
}

impl PlayerSession {
    pub fn new(data: StoredPlayer) -> Self {
        Self {
            data,
            combat_target: None,
            pvp_target: None,
            in_duel: false,
            party_leader: None,
            searched_rooms: HashSet::new(),
            authenticated: false,
        }
    }

    pub fn username(&self) -> &str {
        &self.data.username
    }

    pub fn room_id(&self) -> &str {
        &self.data.room_id
    }

    pub fn in_combat(&self) -> bool {
        self.combat_target.is_some() || self.pvp_target.is_some()
    }

    pub fn clear_combat(&mut self) {
        self.combat_target = None;
        self.pvp_target = None;
    }

    pub fn total_attack(&self) -> i32 {
        let cls = class_stats(self.data.class_name);
        let mut atk = cls.attack + (self.data.level as i32 - 1) * 2;
        if let Some(weapon_id) = &self.data.equipment.weapon {
            if let Some(weapon) = ITEMS.get(weapon_id) {
                if let Some(attack) = weapon.attack {
                    atk += attack;
                }
            }
        }
        atk
    }

    pub fn total_defense(&self) -> i32 {
        let cls = class_stats(self.data.class_name);
        let mut def = cls.defense + (self.data.level as i32 - 1);
        if let Some(armor_id) = &self.data.equipment.armor {
            if let Some(armor) = ITEMS.get(armor_id) {
                if let Some(defense) = armor.defense {
                    def += defense;
                }
            }
        }
        def
    }

    pub fn xp_to_level(&self) -> i32 {
        xp_for_level(self.data.level)
    }

    pub fn add_xp(&mut self, amount: i32) -> Vec<String> {
        let mut messages = Vec::new();
        self.data.xp += amount;
        messages.push(format!("You gain {amount} experience."));

        while self.data.xp >= self.xp_to_level() {
            self.data.xp -= self.xp_to_level();
            self.level_up(&mut messages);
        }
        messages
    }

    fn level_up(&mut self, messages: &mut Vec<String>) {
        self.data.level += 1;
        let cls = class_stats(self.data.class_name);
        let hp_gain = (cls.max_hp as f64 * 0.15).floor() as i32 + 5;
        let mp_gain = (cls.max_mp as f64 * 0.12).floor() as i32 + 3;
        self.data.max_hp += hp_gain;
        self.data.max_mp += mp_gain;
        self.data.hp = self.data.max_hp;
        self.data.mp = self.data.max_mp;
        messages.push(format!(
            "*** LEVEL UP! You are now level {}! ***",
            self.data.level
        ));
        messages.push(format!("HP +{hp_gain}, MP +{mp_gain}"));
    }

    pub fn add_item(&mut self, item_id: &str) {
        self.data.inventory.push(item_id.to_string());
    }

    pub fn remove_item(&mut self, item_id: &str) -> bool {
        if let Some(idx) = self
            .data
            .inventory
            .iter()
            .position(|id| id == item_id)
        {
            self.data.inventory.remove(idx);
            true
        } else {
            false
        }
    }

    pub fn count_item(&self, item_id: &str) -> i32 {
        self.data
            .inventory
            .iter()
            .filter(|id| id.as_str() == item_id)
            .count() as i32
    }

    pub fn get_active_quests(&self) -> Vec<&QuestProgress> {
        self.data
            .quests
            .iter()
            .filter(|q| q.status == QuestStatus::Active)
            .collect()
    }

    pub fn to_snapshot(&self, world: &World) -> PlayerSnapshot {
        let rooms = world.rooms();
        let room = rooms.get(&self.data.room_id);
        PlayerSnapshot {
            username: self.data.username.clone(),
            class_name: self.data.class_name,
            level: self.data.level,
            hp: self.data.hp,
            max_hp: self.data.max_hp,
            mp: self.data.mp,
            max_mp: self.data.max_mp,
            xp: self.data.xp,
            xp_to_level: self.xp_to_level(),
            gold: self.data.gold,
            room: self.data.room_id.clone(),
            room_name: room.map(|r| r.name.clone()),
            zone: room.map(|r| r.zone.clone()),
            in_combat: if self.in_combat() { Some(true) } else { None },
            party_leader: self.party_leader.clone(),
            in_duel: if self.in_duel { Some(true) } else { None },
            title: self.data.title.clone(),
            guild_name: find_guild_by_member(&self.data.username).map(|g| g.name),
        }
    }
}